[RFC] cmake: Add `clint` target to build Makefile

ZyX-I opened the pull request:
Allows linting only modified files and linting multiple files in parallel.

This change is not good for travis: if multiple files have lint errors you may find that only one of them failed build, but good for regular development (especially after I change code to use new files from neovim/bot-ci#94, which will also shrink download times should I use tar.gz).

ZyX-I committed:
cmake: Add `clint` target to build Makefile

Allows linting only modified files and linting multiple files in
parallel. In the current state is rather slow because errors.json is
a 6 MiB file and needs to be reparsed each time.

Results on my system (6-core):

    # In build dir, actually parallel
    make -j5 clint  241.24s user 8.39s system 334% cpu 1:14.74 total
    # In root, one process
    make -j5 clint  60.69s user 0.37s system 93% cpu 1:05.19 total

In both cases download time included.

That is not well for travis (though I would keep travis as-is because
new variant will fail before checking all files), but already good
enough for regular development: total times are nearly identical and
this is the *full* build, further `make -C build clint` will check only
modified files.

ZyX-I committed:
cmake: Use file-specific supprresses

`make -C build clint` time is now

    make -j5 clint  95.29s user 1.86s system 409% cpu 23.751 total

*without* downloading anything (much worse if something was not cached, still 
a bit better then top-level `make clint`). But since without neovim/bot-ci#95 it 
is downloading each file one-by-one total time with download (download also 
parallel!) is

    make -j5 -B clint  99.29s user 2.98s system 258% cpu 39.634 total

Top-level makefile still gives

    make -j5 clint  59.33s user 0.28s system 95% cpu 1:02.41 total

ZyX-I committed:
cmake: Make Download.cmake check for errors

Copying from third-party/cmake/DownloadAndExtractFile.cmake.

ZyX-I commented:
Looks like the result is already better then top-level `make clint`, even though I have to download a bunch of small files. Though current variant will not tolerate new files, need to ignore download failures.

ZyX-I committed:
cmake: Allow failing to download small suppress files

ZyX-I committed:
cmake: Replace RunLint.cmake with code in src/nvim/CMakeLists.txt

This also removes LINT_FILE environment variable, other then that functionality 
is kept. It is expected that developers needing partial linting will use `make 
lint`, touching interesting file before (if not done already by writing to 
them).

ZyX-I commented:
Can be merged without waiting for neovim/bot-ci#95, though it is better to wait and adjust code to use the archive.

bfredl commented:
A drawback is that output is a lot more noisy with a lot of 

    [460/488] Generating ../../touches/ran-clint-screen-h
    Total errors found: 0

lines between the actual errors. Can these be supressed?
Also `LINT_FILE` is not working anymore, but I guess that's intentional/acceptable (as it is not needed for repeated linting anymore).

ZyX-I commented:
LINT_FILE case is explained in the commit I removed it. I do not know a way to remove additional noise without removing something needed (CMake adds some control over this), and actually would not want to: now instead of “`make clint` … wait … did not it go into infinite cycle? … finished” I get information about what is exactly checked now. May modify `clint.py` to remove `Total errors found: 0`, that is purely useless.

ZyX-I committed:
clint: Do not report zero errors

ZyX-I committed:
cmake: Also depend on LINT_PRG

bfredl commented:
> May modify clint.py to remove Total errors found: 0, that is purely useless.

That will be enough to make cmake reuse the same "generating" line for files without errors, and thus solve the noise issue :+1:

justinmk commented on Makefile:
```diff
@@ -126,15 +126,15 @@ distclean: clean
 	+$(BUILD_CMD) -C build check-single-includes
 
-lint: check-single-includes clint testlint
+lint: check-single-includes clint-full testlint
```
Most developers will run `make lint`, and don't want to know about the minutiae of all the various targets. 

Developers who want `make clint-full` can run that directly, it's the uncommon case.

ZyX-I replied:
This target is used in CI scripts. I do not want `make clint` on CI, this may not check all files.

And for “most” some survey is needed. I know for sure I was not ever using that, not even on VMs where I succeeded to use top-level makefile to install dependencies.

justinmk replied:
CI can continue to use `make lint`, it uses a fresh workspace anyway.

"make lint" is just more intuitive, that's why most will use it. We want things to be easy for contributors to use.

ZyX-I replied:
`make lint` already does expensive `check-single-includes`. You do not run it if you care about time.

And by “may not check all files” I did *not* mean that it is going to check only new ones. I meant that once file X5 is checked and script determined that there are errors file X10 scheduled to be run 5 jobs next and also containing errors may be not checked at all because X5 already failed the build. This means developers will see less errors in CI.

ZyX-I replied:
Though `check-single-includes` is a bit faster then difference between `clint` and `clint-full`, and it is also incremental.

justinmk replied:
`check-single-includes` result is cached, just like the new `clint`. So `make lint` can be hammered by developer, it's more convenient.

We can run `clint-full` on CI.

ZyX-I commented on src/nvim/CMakeLists.txt:
```diff
@@ -34,11 +35,19 @@ set(UNICODE_TABLES_GENERATOR ${PROJECT_SOURCE_DIR}/scripts/genunicodetables.lua)
+set(LINT_SUPPRESS_URL_BASE "https://raw.githubusercontent.com/neovim/doc/gh-pages/reports/clint")
+set(LINT_SUPPRESS_URL "${LINT_SUPPRESS_URL_BASE}/errors.json")
+set(LINT_PRG ${PROJECT_SOURCE_DIR}/src/clint.py)
+set(DOWNLOAD_SCRIPT ${PROJECT_SOURCE_DIR}/cmake/Download.cmake)
```
There are at least two occurrences.

ZyX-I committed:
ci: Refactor CI scripts

1. CI_TARGET now determines which run_${CI_TARGET}.sh script to use. Defaults to
   `tests`.
2. Build no longer halts on the first failing suit: e.g. if functional tests
   failed it will continue with unit tests, etc.
3. All ${MAKE_CMD} occurrences moved to `top_make` function, added `build_make`
   as an alias to `make -C build` (`"${MAKE_CMD}" -C "${BUILD_DIR}"`) which is
   too verbose.

`suite.sh` was copied from powerline (tests/common.sh file), assumes running
with POSIX shells (and actually uses dash in powerline). Then some convenience
functions were added (run_test and below).

ZyX-I committed:
makefile: Make `lint` target depend on `clint`

ZyX-I commented:
Remove another thing which also bothered me: tests do not proceed on failure. E.g. you may fix clint only to find out that testlint fails now. New variant always proceeds with the next tests, failing at the end of the script.

ZyX-I commented:
This has one downside though: it may appear that failure reason is in the middle of the log. Not much of a problem with unit and functional tests (it *always* is in the middle), lint is worse here.

ZyX-I committed:
ci: Do not quote MAKE_CMD

ZyX-I committed:
ci: Add ${NL} variable

jamessan commented on ci/common/suite.sh:
```diff
@@ -1,3 +1,7 @@
+# HACK: get newline for use in strings given that "\n" and $'' do not work.
+NL="$(printf '\nE')"
```
Can't you just do this?

```sh
NL="
"
```

ZyX-I replied:
This works only as long as you run it on the zero level of indentation. Move NL definition to some function and this is no longer an option.

ZyX-I committed:
ci: Check for exact value of CI_TARGET, not its emptyness

ZyX-I committed:
ci: Enable tracing

ZyX-I committed:
ci: Add test watchdog and tracing for lint tests

ZyX-I committed:
ci: Do not shift if there are not enough arguments

justinmk commented on ci/common/suite.sh:
```diff
@@ -45,23 +45,23 @@ fail() {
   local cmd="$1"
-  shift
+  test $# -gt 0 && shift
   local test_name="$1"
```
If `test $# -gt 0 && shift` fails won't `test_name` have the same value as `$cmd`? Why not use `$2` instead of  `shift`?

Also the file is missing shebang and error options that we usually set in bash scripts.

ZyX-I replied:
shebang and error options are *never* set in ci/common scripts.

ZyX-I replied:
And repetive `test $# -gt 0 && shift` will shift up until no arguments left. So test_name will have the same value only if first argument was missing, and further code handles empty `$1`.

ZyX-I replied:
`$2`, etc cannot be used because arguments used by this function must be got rid of so that `$@` passed to `fail` will have expected arguments.
